File: src/domains/ads-stats.ts
import { mkdirSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { stringify } from "csv-stringify/sync";

export interface Deps {
  graphApi: (
    method: string,
    endpoint: string,
    token: string,
    params?: Record<string, string>,
    body?: Record<string, unknown>,
  ) => Promise<any>;
}

export interface StatsPoint {
  campaign_id: string;
  campaign_name?: string;
  date: string;
  impressions: number;
  clicks: number;
  spend: number;
  ctr: number;
  cpc: number;
  cpm: number;
  conversions: number;
  conversion_value: number;
  cpa: number;
  roas: number;
}

function toNumber(value: unknown): number {
  if (typeof value === "number") return value;
  const parsed = Number(value ?? 0);
  return Number.isFinite(parsed) ? parsed : 0;
}

function stddev(values: number[], avg: number): number {
  if (values.length === 0) return 0;
  const variance = values.reduce((sum, v) => sum + (v - avg) ** 2, 0) / values.length;
  return Math.sqrt(variance);
}

function summarize(values: number[]) {
  if (values.length === 0) return { min: 0, max: 0, avg: 0, stddev: 0 };
  const min = Math.min(...values);
  const max = Math.max(...values);
  const avg = values.reduce((a, b) => a + b, 0) / values.length;
  return { min, max, avg, stddev: stddev(values, avg) };
}

function normalizeAccountPath(accountId: string): string {
  return accountId.startsWith("act_") ? accountId : `act_${accountId}`;
}

function parseInsightsRow(row: Record<string, unknown>): StatsPoint {
  const conversions = toNumber(row.conversions);
  const conversionValue = toNumber(row.conversion_value);
  const spend = toNumber(row.spend);
  return {
    campaign_id: String(row.campaign_id ?? ""),
    campaign_name: row.campaign_name ? String(row.campaign_name) : undefined,
    date: String(row.date_start ?? ""),
    impressions: toNumber(row.impressions),
    clicks: toNumber(row.clicks),
    spend,
    ctr: toNumber(row.ctr),
    cpc: toNumber(row.cpc),
    cpm: toNumber(row.cpm),
    conversions,
    conversion_value: conversionValue,
    cpa: conversions > 0 ? spend / conversions : 0,
    roas: spend > 0 ? conversionValue / spend : 0,
  };
}

export async function collectStats(
  deps: Deps,
  accountId: string,
  token: string,
  startDate: string,
  endDate: string,
  storageDir = `${process.env.HOME ?? "~"}/.config/facebook-cli/stats/daily`,
) {
  mkdirSync(storageDir, { recursive: true });

  const accountPath = normalizeAccountPath(accountId);
  const response = await deps.graphApi("GET", `${accountPath}/insights`, token, {
    level: "campaign",
    time_increment: "1",
    fields: "campaign_id,campaign_name,impressions,clicks,spend,ctr,cpc,cpm,date_start,date_stop,conversions,conversion_value",
    time_range: JSON.stringify({ since: startDate, until: endDate }),
  });

  const rows: Array<Record<string, unknown>> = Array.isArray(response?.data) ? response.data : [];
  const points = rows.map(parseInsightsRow);

  for (const point of points) {
    const filename = `${point.campaign_id}_${point.date}.json`;
    writeFileSync(join(storageDir, filename), JSON.stringify(point, null, 2));
  }

  return {
    campaigns: new Set(points.map((p) => p.campaign_id)).size,
    dataPoints: points.length,
    storageDir,
    points,
  };
}

export function analyzeStats(dataPoints: Array<Record<string, unknown>>) {
  const impressions = dataPoints.map((d) => toNumber(d.impressions));
  const clicks = dataPoints.map((d) => toNumber(d.clicks));
  const spend = dataPoints.map((d) => toNumber(d.spend));
  const ctr = dataPoints.map((d) => toNumber(d.ctr));
  const cpc = dataPoints.map((d) => toNumber(d.cpc));
  const cpm = dataPoints.map((d) => toNumber(d.cpm));

  return {
    impressions: summarize(impressions),
    clicks: summarize(clicks),
    spend: summarize(spend),
    ctr: summarize(ctr),
    cpc: summarize(cpc),
    cpm: summarize(cpm),
    trend: {
      impressions: impressions.length > 1 ? impressions[impressions.length - 1] - impressions[0] : 0,
      clicks: clicks.length > 1 ? clicks[clicks.length - 1] - clicks[0] : 0,
      spend: spend.length > 1 ? spend[spend.length - 1] - spend[0] : 0,
    },
  };
}

export function validateStats(dataPoints: StatsPoint[]) {
  const byCampaign = new Map<string, StatsPoint[]>();
  for (const point of dataPoints) {
    const arr = byCampaign.get(point.campaign_id) ?? [];
    arr.push(point);
    byCampaign.set(point.campaign_id, arr);
  }

  return Array.from(byCampaign.entries()).map(([campaignId, points]) => {
    const impressions = points.reduce((sum, p) => sum + p.impressions, 0);
    const clicks = points.reduce((sum, p) => sum + p.clicks, 0);
    const spend = points.reduce((sum, p) => sum + p.spend, 0);
    const runtimeHours = points.length * 24;
    const pass = impressions >= 1000 && clicks >= 10 && spend >= 1 && runtimeHours >= 24;
    return {
      campaignId,
      pass,
      checks: {
        impressions,
        clicks,
        spend,
        runtimeHours,
      },
      recommendation: pass
        ? "ready_for_optimization"
        : "collect_more_data_before_optimization",
    };
  });
}

export function exportStatsCsv(dataPoints: StatsPoint[], outputPath: string) {
  const csv = stringify(
    dataPoints.map((d) => ({
      campaign_id: d.campaign_id,
      campaign_name: d.campaign_name ?? "",
      date: d.date,
      impressions: d.impressions,
      clicks: d.clicks,
      spend: d.spend,
      ctr: d.ctr,
      cpc: d.cpc,
      cpm: d.cpm,
      conversions: d.conversions,
      cpa: d.cpa,
      roas: d.roas,
    })),
    { header: true },
  );
  writeFileSync(outputPath, csv);
}

File: internal/facebook/ads_stats.go
package facebook

import (
	"context"
	"encoding/csv"
	"encoding/json"
	"math"
	"os"
	"path/filepath"
	"strconv"

	"github.com/salmonumbrella/facebook-cli/internal/api"
)

type StatsPoint struct {
	CampaignID      string  `json:"campaign_id"`
	CampaignName    string  `json:"campaign_name,omitempty"`
	Date            string  `json:"date"`
	Impressions     float64 `json:"impressions"`
	Clicks          float64 `json:"clicks"`
	Spend           float64 `json:"spend"`
	CTR             float64 `json:"ctr"`
	CPC             float64 `json:"cpc"`
	CPM             float64 `json:"cpm"`
	Conversions     float64 `json:"conversions"`
	ConversionValue float64 `json:"conversion_value"`
	CPA             float64 `json:"cpa"`
	ROAS            float64 `json:"roas"`
}

func toNumber(value any) float64 {
	switch typed := value.(type) {
	case float64:
		return typed
	case float32:
		return float64(typed)
	case int:
		return float64(typed)
	case int64:
		return float64(typed)
	case string:
		number, err := strconv.ParseFloat(typed, 64)
		if err == nil {
			return number
		}
		return 0
	case json.Number:
		number, _ := typed.Float64()
		return number
	default:
		return 0
	}
}

func parseInsightsRow(row map[string]any) StatsPoint {
	conversions := toNumber(row["conversions"])
	conversionValue := toNumber(row["conversion_value"])
	spend := toNumber(row["spend"])
	point := StatsPoint{
		CampaignID:      asString(row["campaign_id"]),
		CampaignName:    asString(row["campaign_name"]),
		Date:            asString(row["date_start"]),
		Impressions:     toNumber(row["impressions"]),
		Clicks:          toNumber(row["clicks"]),
		Spend:           spend,
		CTR:             toNumber(row["ctr"]),
		CPC:             toNumber(row["cpc"]),
		CPM:             toNumber(row["cpm"]),
		Conversions:     conversions,
		ConversionValue: conversionValue,
	}
	if conversions > 0 {
		point.CPA = spend / conversions
	}
	if spend > 0 {
		point.ROAS = conversionValue / spend
	}
	return point
}

func CollectStats(
	ctx context.Context,
	client *api.Client,
	accountID string,
	token string,
	startDate string,
	endDate string,
	storageDir string,
) (map[string]any, error) {
	if storageDir == "" {
		home, _ := os.UserHomeDir()
		storageDir = filepath.Join(home, ".config", "facebook-cli", "stats", "daily")
	}
	if err := os.MkdirAll(storageDir, 0o755); err != nil {
		return nil, err
	}

	response, err := client.Graph(ctx, "GET", normalizeAccountPath(accountID)+"/insights", token, map[string]string{
		"level":          "campaign",
		"time_increment": "1",
		"fields":         "campaign_id,campaign_name,impressions,clicks,spend,ctr,cpc,cpm,date_start,date_stop,conversions,conversion_value",
		"time_range":     normalizeValue(map[string]any{"since": startDate, "until": endDate}),
	}, nil)
	if err != nil {
		return nil, err
	}
	row := normalizeGraphObject(response)
	items, _ := row["data"].([]any)
	points := make([]StatsPoint, 0, len(items))
	seenCampaigns := map[string]struct{}{}
	for _, item := range items {
		point := parseInsightsRow(normalizeGraphObject(item))
		points = append(points, point)
		seenCampaigns[point.CampaignID] = struct{}{}

		filename := filepath.Join(storageDir, point.CampaignID+"_"+point.Date+".json")
		data, err := json.MarshalIndent(point, "", "  ")
		if err != nil {
			return nil, err
		}
		if err := os.WriteFile(filename, data, 0o644); err != nil {
			return nil, err
		}
	}

	return map[string]any{
		"campaigns":  len(seenCampaigns),
		"dataPoints": len(points),
		"storageDir": storageDir,
		"points":     points,
	}, nil
}

func summarize(values []float64) map[string]any {
	if len(values) == 0 {
		return map[string]any{"min": 0.0, "max": 0.0, "avg": 0.0, "stddev": 0.0}
	}
	minValue := values[0]
	maxValue := values[0]
	total := 0.0
	for _, value := range values {
		if value < minValue {
			minValue = value
		}
		if value > maxValue {
			maxValue = value
		}
		total += value
	}
	avg := total / float64(len(values))
	var variance float64
	for _, value := range values {
		diff := value - avg
		variance += diff * diff
	}
	variance /= float64(len(values))
	return map[string]any{
		"min":    minValue,
		"max":    maxValue,
		"avg":    avg,
		"stddev": math.Sqrt(variance),
	}
}

func AnalyzeStats(dataPoints []map[string]any) map[string]any {
	impressions := []float64{}
	clicks := []float64{}
	spend := []float64{}
	ctr := []float64{}
	cpc := []float64{}
	cpm := []float64{}

	for _, point := range dataPoints {
		impressions = append(impressions, toNumber(point["impressions"]))
		clicks = append(clicks, toNumber(point["clicks"]))
		spend = append(spend, toNumber(point["spend"]))
		ctr = append(ctr, toNumber(point["ctr"]))
		cpc = append(cpc, toNumber(point["cpc"]))
		cpm = append(cpm, toNumber(point["cpm"]))
	}

	trend := map[string]any{"impressions": 0.0, "clicks": 0.0, "spend": 0.0}
	if len(impressions) > 1 {
		trend["impressions"] = impressions[len(impressions)-1] - impressions[0]
		trend["clicks"] = clicks[len(clicks)-1] - clicks[0]
		trend["spend"] = spend[len(spend)-1] - spend[0]
	}

	return map[string]any{
		"impressions": summarize(impressions),
		"clicks":      summarize(clicks),
		"spend":       summarize(spend),
		"ctr":         summarize(ctr),
		"cpc":         summarize(cpc),
		"cpm":         summarize(cpm),
		"trend":       trend,
	}
}

func ValidateStats(dataPoints []StatsPoint) []map[string]any {
	byCampaign := map[string][]StatsPoint{}
	for _, point := range dataPoints {
		byCampaign[point.CampaignID] = append(byCampaign[point.CampaignID], point)
	}

	out := make([]map[string]any, 0, len(byCampaign))
	for campaignID, points := range byCampaign {
		impressions := 0.0
		clicks := 0.0
		spend := 0.0
		for _, point := range points {
			impressions += point.Impressions
			clicks += point.Clicks
			spend += point.Spend
		}
		runtimeHours := len(points) * 24
		pass := impressions >= 1000 && clicks >= 10 && spend >= 1 && runtimeHours >= 24
		recommendation := "collect_more_data_before_optimization"
		if pass {
			recommendation = "ready_for_optimization"
		}
		out = append(out, map[string]any{
			"campaignId": campaignID,
			"pass":       pass,
			"checks": map[string]any{
				"impressions":  impressions,
				"clicks":       clicks,
				"spend":        spend,
				"runtimeHours": runtimeHours,
			},
			"recommendation": recommendation,
		})
	}
	return out
}

func ExportStatsCSV(dataPoints []StatsPoint, outputPath string) error {
	file, err := os.Create(outputPath)
	if err != nil {
		return err
	}
	defer func() { _ = file.Close() }()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	if err := writer.Write([]string{
		"campaign_id", "campaign_name", "date", "impressions", "clicks", "spend", "ctr", "cpc", "cpm", "conversions", "cpa", "roas",
	}); err != nil {
		return err
	}
	for _, point := range dataPoints {
		if err := writer.Write([]string{
			point.CampaignID,
			point.CampaignName,
			point.Date,
			asString(point.Impressions),
			asString(point.Clicks),
			asString(point.Spend),
			asString(point.CTR),
			asString(point.CPC),
			asString(point.CPM),
			asString(point.Conversions),
			asString(point.CPA),
			asString(point.ROAS),
		}); err != nil {
			return err
		}
	}
	return writer.Error()
}
